### PR TITLE
Add model and logic to create a new ChargeLink

### PR DESCRIPTION
**Pull request by BjarkeMeier**
## Description

Add model and logic to create a new ChargeLink.

Note: Does not include unit tests and persistence.

## References

* #334

### CODE REVIEW

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/ChargeLinks/ChargeLinkOperation.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        private const int MaxIdLength = 100;
+        private const int RequiredCorrelationIdLength = 36;
+
+        public ChargeLinkOperation(string id, string correlationId)
+        {
+            if (id == null) throw new ArgumentNullException(nameof(id));
+            if (id.Length > MaxIdLength) throw new ArgumentException($"Must not exceed {MaxIdLength} characters.", nameof(id));
+
+            if (correlationId == null) throw new ArgumentNullException(nameof(correlationId));
+            if (correlationId.Length > RequiredCorrelationIdLength) throw new ArgumentException($"Must be {RequiredCorrelationIdLength} characters long.", nameof(correlationId));
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/ChargeLinks/ChargeLinkOperation.cs:31]: I think this needs another look. Do we want to ensure that exactly 36 is used, or do we want to ensure a maximum of 36 is used?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Context/ChargeLinkDatabaseContext.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+            {
+                builder.ToTable("ChargeLink");
+                //builder.OwnsMany(cl => cl.)
+            });
+
+            //modelBuilder.Entity<ChargeOperation>().ToTable("ChargeOperation");
+            //modelBuilder.Entity<ChargePeriodDetails>().ToTable("ChargePeriodDetails");
+            //modelBuilder.Entity<Charge>().ToTable("Charge");
+            //modelBuilder.Entity<MarketParticipant>().ToTable("MarketParticipant");
+            //modelBuilder.Entity<MeteringPoint>().ToTable("MeteringPoint");
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Context/ChargeLinkDatabaseContext.cs:53]: Is this here for a reason?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/Charge.cs [hunk 3]
```diff
@@ -35,7 +40,7 @@ public Charge()
         public string ChargeOperationId { get; set; }
 
         /// <summary>
-        /// Unique ID of a charge (Note, unique per market participants).
+        /// Unique ID of a charge (Note, unique per market participants AND charge type).
```
Comment (Mech0z) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/Charge.cs:43]: lower case "and" ?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/MeteringPoints/MeteringPoint.cs [hunk 4]
```diff
@@ -34,6 +34,8 @@ public class MeteringPoint
             SettlementMethod = settlementMethod;
         }
 
+        public int MeteringPointRowId { get; private set; }
```
Comment (Mech0z) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/MeteringPoints/MeteringPoint.cs:37]: remove private set

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Context/ChargeLinkDatabaseContext.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+using System;
+using System.Threading.Tasks;
+using GreenEnergyHub.Charges.Domain.ChargeLinks;
+using GreenEnergyHub.Charges.Domain.ChargeLinks.Command;
+using Microsoft.EntityFrameworkCore;
+
+namespace GreenEnergyHub.Charges.Infrastructure.Context
+{
+    public class ChargeLinkDatabaseContext : DbContext, IChargeLinkDatabaseContext
+    {
```
Comment (Mech0z) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Context/ChargeLinkDatabaseContext.cs:24]: Currently we only have a single context file for Charges and Metering Points, if we want a separate for Charge link, should we also refactor Metering Points to have its own context or move this context to Charges context?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/ChargeLinks/ChargeLinkOperation.cs (hunk 1 above)
Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/ChargeLinks/ChargeLinkOperation.cs]: I'm unsure. But I've loosened up.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Context/ChargeLinkDatabaseContext.cs (hunk 2 above)
Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Context/ChargeLinkDatabaseContext.cs]: I had started working on next PR. Removed.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/Charge.cs (hunk 3 above)
Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Charges/Charge.cs]: Fixed.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/MeteringPoints/MeteringPoint.cs (hunk 4 above)
Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/MeteringPoints/MeteringPoint.cs:37]: And add ctor arg?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Context/ChargeLinkDatabaseContext.cs (hunk 5 above)
Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Context/ChargeLinkDatabaseContext.cs:23]: What do you recommend? I don't have an opinion as of now - but I most probably will have during my next PR when implementing persistence.

Comment (Mech0z) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Context/ChargeLinkDatabaseContext.cs:23]: With our 3 small subdomains I think its easier to have them in the same context, also if we know we need to do joins later I am not certain how different contexts work together

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/MeteringPoints/MeteringPoint.cs (hunk 4 above)
Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/MeteringPoints/MeteringPoint.cs]: Fixed.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Context/ChargeLinkDatabaseContext.cs (hunk 5 above)
Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/Context/ChargeLinkDatabaseContext.cs]: Ok, removed.

**Review by Mech0z — APPROVED** [review]

**Pull request merged**